middle_end/flambda/to_cmm/un_cps_closure.ml

```diff
@@ -105,40 +113,84 @@ let order_env_vars env l acc =
             Var_within_closure.print env_var
     ) acc l
 
-let order env closures env_vars =
-  order_env_vars env env_vars
-    (order_closures env closures Numbers.Int.Map.empty)
-
-let rec layout_aux map acc i =
-  match Numbers.Int.Map.find_first (fun j -> j >= i) map with
-  | exception Not_found -> List.rev acc
-  | (j, (Closure _ as slot)) when j = 0 ->
-      assert (i = j);
-      assert (acc = []);
-      layout_aux map [(0, slot)] (i + 1)
-  | (j, (Closure _ as slot)) ->
-      layout_aux map ((j, slot) :: (j - 1, Infix_header) :: acc) (j + 1)
-  | (j, (Env_var _ as slot)) ->
-      layout_aux map ((j, slot) :: acc) (j + 1)
-  | (_, Infix_header) ->
+let layout_aux j slot (startenv, acc_slots) =
+  match slot with
+  (* Starting from ocaml 4.12, all closure slots *must* precede all env var
+     slots. The algorithms in this file should thus only generate slot
+     assignments that respect this invariant. If that is not the case, this
+     is a fatal error given that the start of the environment (i.e. the
+     offset of the first en var slot, with the added property the all slots
+     after that are env vars slots (or at least scannable by the CG), is needed
```

Comment:
```suggestion
     offset of the first env var slot, with the added property that all slots
     after that are env vars slots (or at least scannable by the CG)), is needed
```

Reply:
There is now a duplicate parenthesis...

Reply:
There was a parenthesis inside a parenthesis, and only one got closed, so I added a closing parenthesis where I thought it was intended

Reply:
Actually, it's normal because the text was already inside another parenthesis (but the first opening parenthesis is not shown on the small diff shown in the comments).